### PR TITLE
add BitBox02 hardware wallet support

### LINKED ISSUE #8119: BitBox02 hardware wallet integration
**What problem are you trying to solve?**

I am a developer at Shift Cryptosecurity (https://shiftcrypto.ch) and would like to integrate our BitBox02 HWW into Metamask. 
We (Shift Cryptosecurity) have been around since 2015 and we have had two successful products in this space: the "Digital BitBox (BitBox01)" and the "BitBox02".

The BitBox01 has been integrated in MyEtherWallet since 2017 and the BitBox02 integration is an open PR on MEW.

I talked to @brunobar79 about this recently and he suggested to open an issue to discuss details first. He pointed me to the trezor and ledger keyring libraries that he made to integrate Ledger and Trezor. We have a javascript library for our device here: https://www.npmjs.com/package/bitbox02-api and from looking at your keyrings, I should be able to easily provide you with one of those for our device. I see that Bruno is still the maintainer for the ledger and trezor keyring packages on npm so my question is if you would like to develop the keyring yourself, or if I provide you with one, who should maintain the package. 
 
 **Describe the solution you'd like**

The integration of our hww into services such as Metamask or MEW is quite unique as it does not use u2f hijack, webusb, or an external website/popup that the user has to worry about. Our goal is that when using the BB02, the user stays within the interface of Metamask the entire time. Our BitBoxBridge (https://github.com/digitalbitbox/bitbox-bridge) runs on localhost, whitelists services that want to integrate our device and relays the traffic which is end-to-end encrypted all the way to the browser.  We never redirect the user to anything hosted by us, so we can't collect metadata about the users.

What would be needed for Metamask is to add a few UI elements to guide the user through the process of connecting, unlocking and pairing (verifying the encrypted channel pairing code) the device within the Metamask UI. For an idea what that might look like, I am adding a gif below for what this process looks like on MEW. Please let me know if you would prefer me to do the UI integration, or if this is something you would rather take care of yourself. 

74089043-3a7fdb00-4a9d-11ea-922c-24e33c6959f7

 **Additional context**

You can contact me at tv@shiftcrypto.ch
For testing and development purposes, we can send you a developer version of the BitBox02 which can be loaded with unsigned firmware. You can contact me on my email with shipping details.

We've had our BitBox01 integrated on MyEtherWallet for a few years now and currently have an open PR to integrate the BB02 as well: https://github.com/MyEtherWallet/MyEtherWallet/pull/1957

Thanks!

### PR DESCRIPTION
**Pull request by benma**
Fixes: #8119

[Shift](https://shiftcrypto.ch/) is receiving requests for a BitBox02 Metamask integration nearly on a daily basis.

Previously we waited with working on the integration until the [plugin system](https://github.com/MetaMask/metamask-snaps-beta/wiki) was ready. 

However, since user requests are piling up, we took another look and saw that the integration is actually quite trivial:

- Add BitBox02 next to Trezor/Ledger in the lists for hardware wallets
- Copy the [Trezor keyring](https://github.com/MetaMask/eth-trezor-keyring/blob/c606543f2bbd4135b8bf1a83c45ceaac71b3461c/index.js) to `eth-bitbox02-keyring.js` and fill in the connection for the BitBox02

As a result, the diff is quite minimal and easy to understand and maintain.

When the plugin system is ready, we will be happy to port the integration to that if desired.

We will also be happy to maintain this integration and fix any issues that might arise. We are generally careful not to introduce any breaking changes.

Manual testing steps:  
  - Install the [BitBoxBridge](https://shiftcrypto.ch/download/)
  - Connect an initialized BitBox02 (firmware v9.7.0 or later)
  - Connect it as a hardware wallet and use it to sign transactions and messages

### CODE REVIEW

**Comment by benma** [comment]
I have read the CLA Document and I hereby sign the CLA

**Comment by eerne** [comment]
Wow BitBox support coming for real? That's exciting news 👍  🎉

**Comment by My1** [comment]
> Wow BitBox support coming for real? That's exciting news

well the bitbox people now really did all the work, now metamask just has to push the button

**Comment by benma** [comment]
There are some CI errors that I will get to, but `yarn start` and `yarn dist` should work in the meantime.

**Comment by JohannesHoppe** [comment]
End user here. Please approve this! 😉

File: app/scripts/metamask-controller.js [hunk 1]
```diff
@@ -2120,16 +2137,19 @@ export default class MetamaskController extends EventEmitter {
 
   /**
    * Method to return a boolean if the keyring for the currently selected
-   * account is a ledger or trezor keyring.
-   * TODO: remove this method when Ledger and Trezor release their supported
+   * account is a bitbox02 or ledger or trezor keyring.
+   * TODO: remove this method when BitBox02 and Trezor release their supported
```
Comment (JohannesHoppe) [inline app/scripts/metamask-controller.js:2141]: 😟

File: ui/selectors/selectors.js [hunk 2]
```diff
@@ -87,9 +87,13 @@ export function getCurrentKeyring(state) {
 }
 
 export function isEIP1559Account(state) {
-  // Trezor does not support 1559 at this time
+  // Trezor amd BitBox02 do not support 1559 at this time
```
Comment (JohannesHoppe) [inline ui/selectors/selectors.js:90]:
```suggestion
  // Trezor and BitBox02 do not support 1559 at this time
```

**Comment by devourer66** [comment]
Hi @benma , Who and What needs doing to get this merged?

**Comment by Flash1232** [comment]
Maybe they are just elaborating on how to best integrate new hardware wallets? There's also #12053 that is waiting to get merged.

**Comment by Flash1232** [comment]
I'm fairly certain this PR leads us nowhere, unfortunately, except if the Shift Crypto staff has been reached out to (I remember to read something regarding E-Mail exchange), because https://github.com/MetaMask/metamask-extension/pull/8196#issuecomment-917122953 sums up pretty clearly what's going on.

I would really appreciate the forking route if that's working just for BitBox02. Or are there any uneasy implications with that? I mean I suppose technically we could just fork MetaMask and add the commits from here and upload it as a private Dev plugin to Mozilla/ import in Chrome for us to use. But then it's really frustrating with getting upstream updates and such...

**Comment by benma** [comment]
@danjm @PeterYinusa since you reviewed and approved https://github.com/MetaMask/metamask-extension/pull/12053, is there any chance of taking a look at this PR? We were told to wait for the plugin system to finalize, but since Lattice1 support was added without it, I figured maybe the situation has changed.

**Comment by jacobcantele** [comment]
Hi @benma, as we had previously discussed via email, we have been open to adding QR code based hw due to open standards, and were also adding the Lattice due to its encrypted wifi based connection. USB based hardware wallets and the related APIs are pretty unstable, and these add significant technical debt for us. For other USB-based hw, the path is still Snaps. We'll have some progress to show there very soon.

**Comment by vogues** [comment]
This integration would be wonderful!

**Comment by feld** [comment]
I have just rebased this branch locally and built my own Metamask extension to use on Firefox Dev. The Bitbox02 Metamask integration is far superior to the experience of using my Ledger Nano X. All of my complaints about using hardware wallets and Metamask don't apply when using the Bitbox.

Please, I am begging you... merge this.

**Comment by vogues** [comment]
> I have just rebased this branch locally and built my own Metamask extension to use on Firefox Dev. The Bitbox02 Metamask integration is far superior to the experience of using my Ledger Nano X. All of my complaints about using hardware wallets and Metamask don't apply when using the Bitbox.
> 
> Please, I am begging you... merge this.

Do you need money? Let's get a donation going.

**Comment by My1** [comment]
how about no?

**Comment by schmitmd** [comment]
Still interested in this, please don't close

**Comment by My1** [comment]
would be neat if someone would actually think about adding the bitbox tho

**Comment by schmitmd** [comment]
I heard rumblings that WalletConnect is soon to be added to the BitBox02 natively.  Perhaps that avoids the need for this altogether...?

**Pull request closed without merging**